### PR TITLE
Represent Pubspec.dependencies as a map

### PR DESCRIPTION
**Pull request by nex3**
This cleans up some existing code, but more importantly it makes some
algorithms for the new version solver more efficient.

### CODE REVIEW

**Comment by kevmoo** [comment]
There's a plan for a new solver? I'm so excited!

**Comment by nex3** [comment]
There's not just a plan, there's a working prototype!

File: lib/src/package.dart [hunk 1]
```diff
@@ -49,31 +49,21 @@ class Package {
...
-    addToMap(dep) {
-      deps[dep.name] = dep;
-    }
-
-    dependencies.forEach(addToMap);
-    devDependencies.forEach(addToMap);
-
+  Map<String, PackageRange> get immediateDependencies {
+    return {}..addAll(dependencies)..addAll(devDependencies)
     // Make sure to add these last so they replace normal dependencies.
```
Comment (munificent) [inline lib/src/package.dart]: Putting the comment in the middle of the cascade is kind of weird. How about moving it up and changing "these" -> "overrides"?

File: lib/src/pubspec.dart [hunk 2]
```diff
@@ -597,11 +597,11 @@ class Pubspec {
...
       SourceRegistry sources})
       : _version = version,
-        _dependencies = dependencies == null ? null : dependencies.toList(),
+        _dependencies = dependencies == null ? null : new Map.fromIterable(dependencies, key: (range) => range.name),
         _devDependencies =
-            devDependencies == null ? null : devDependencies.toList(),
+            devDependencies == null ? null : new Map.fromIterable(devDependencies, key: (range) => range.name),
         _dependencyOverrides =
-            dependencyOverrides == null ? null : dependencyOverrides.toList(),
+            dependencyOverrides == null ? null : new Map.fromIterable(dependencyOverrides, key: (range) => range.name),
```
Comment (munificent) [inline lib/src/pubspec.dart]: Dartfmt?

File: lib/src/package.dart (hunk 1 above)
Comment (nex3) [reply lib/src/package.dart]: Done.

File: lib/src/pubspec.dart (hunk 2 above)
Comment (nex3) [reply lib/src/pubspec.dart]: Done.

**Pull request merged**